src/runner.ts

```diff
@@ -1,3 +1,112 @@
+import * as dopri from "dopri";
+import type { UserType } from "./user";
+import * as userHelpers from "./user";
+
+// Probably this is something that dopri should export for us, we
+// could also use its types for the rhs and output members below.
+type Solution = (t: number) => number[];
+
+export type OdinModelConstructable =
+    new(userHelpers: any, pars: UserType, unknownAction: string) => OdinModel;
+
+interface OdinModelODE {
+    initial(t: number): number[];
+    rhs(t: number, y: number[], dydt: number[]): void;
+    output?(t: number, y: number[]): number[];
+    names(): string[];
+}
+
+interface OdinModelDDE {
+    initial(t: number): number[];
+    rhs(t: number, y: number[], dydt: number[], solution: Solution): void;
+    output?(t: number, y: number[], solution: Solution): number[];
+    names(): string[];
+}
+
+function isDDEModel(model: OdinModel): model is OdinModelDDE {
+    return model.rhs.length === 4;
+}
+
+export type OdinModel = OdinModelODE | OdinModelDDE;
+
+export function delay(solution: Solution, t: number, index: number[],
+                      state: number[]) {
+    // Later, we'll update dopri.js to allow passing index here,
+    // which will make this more efficient. However, no change to
+    // the external interface will be neeed.
+    const y = solution(t);
+    for (let i = 0; i < index.length; ++i) {
+        state[i] = y[index[i]];
+    }
+}
+
+export const base = {checkUser: userHelpers.checkUser,
+                     delay,
+                     getUserScalar: userHelpers.getUserScalar};
+
+// tslint:disable-next-line:variable-name
+export function wodinRun(Model: OdinModelConstructable, pars: UserType,
+                         tStart: number, tEnd: number,
+                         control: any) {
+    const model = new Model(base, pars, "error");
+    const solution = (
+        isDDEModel(model) ?
+            wodinRunDDE(model as OdinModelDDE, tStart, tEnd, control) :
+            wodinRunODE(model as OdinModelODE, tStart, tEnd, control));
+    const names = model.names();
+    return (t0: number, t1: number, nPoints: number) => {
+        const t = grid(Math.max(0, t0), Math.min(t1, tEnd), nPoints);
+        const y = solution(t);
+        return y[0].map((_: any, i: number) => ({
+            name: names[i], x: t, y: y.map((row: number[]) => row[i])}));
+    };
+}
+
+function wodinRunODE(model: OdinModelODE, tStart: number, tEnd: number,
+                     control: any) {
+    // tslint:disable-next-line:only-arrow-functions
+    const rhs = function(t: number, y: number[], dydt: number[]) {
+        model.rhs(t, y, dydt);
+    };
+
+    let output = null;
+    if (typeof model.output === "function") {
+        // Without 'as' here, TS thinks that model.output could be
+        // rebound and no longer a function. We tried saving
+        //   const output = model.output;
+        // which pleases the compiler but then fails at runtime.
+        // tslint:disable-next-line:ban-types
```

Comment:
This is the fix for the thing I was alluding to in standup - our previous trick of saving `model.output` lead to runtime failures in models that had output functions that use internals (for example, the DDE example)

Reply:
Because it might be undefined? Would this not work:
```
if (model.output) {
  output = (t: number, y: number[]) => model.output(t, y);
}
```

> runtime failures in models that had output functions that use internals

not sure I understand that, maybe we could chat about this tomorrow?

Reply:
That's very werid. No idea! :(

Reply:
"that use internals" - so that's because the `this` keyword isn't bound correctly after `model.output` is copied into a local variable?

Reply:
I wonder if `let output = model.output; output = output.bind(model)` might work

Reply:
It might, but I think the current workaround will probably cause less headscratching? I suspect what we should do is have 4 interfaces rather than `output?` so that the model is _always_ of the same type